go/cmd/master/master.go

```diff
@@ -5,41 +5,80 @@ import (
 	"net/http"
 	"net/rpc"
 	"strconv"
+	"strings"
+	"sync"
 	"time"
 
 	"github.com/namsral/flag"
+	log "github.com/sirupsen/logrus"
 
 	"github.com/PaddlePaddle/Paddle/go/master"
 )
 
+type inMemStore struct {
+	mu  sync.Mutex
+	buf []byte
+}
+
+func (m *inMemStore) Save(b []byte) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.buf = b
+	return nil
+}
+
+func (m *inMemStore) Load() ([]byte, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	return m.buf, nil
+}
+
 func main() {
 	port := flag.Int("port", 8080, "port of the master server.")
 
-	faultTolerance := flag.Bool("fault_tolerance", false, "enable fault tolerance (requires etcd).")
+	ttlSec := flag.Int("ttl", 60, "etcd lease TTL in seconds.")
+	endpoints := flag.String("endpoints", "", "comma separated etcd endpoints. If empty, fault tolerance will not be enabled.")
 	taskTimeoutDur := flag.Duration("task_timout_dur", 20*time.Minute, "task timout duration.")
 	taskTimeoutMax := flag.Int("task_timeout_max", 3, "max timtout count for each task before it being declared failed task.")
 	chunkPerTask := flag.Int("chunk_per_task", 10, "chunk per task.")
 	flag.Parse()
 
-	if *faultTolerance {
-		panic("fault tolernance not implemented.")
+	if *endpoints == "" {
+		log.Warningln("-endpoints not set, fault tolerance not be enabled.")
+	}
+
+	var store master.Store
+	if *endpoints != "" {
+		eps := strings.Split(*endpoints, ",")
+		var err error
+		store, err = master.NewEtcdStore(eps, master.DefaultLockPath, master.DefaultStatePath, *ttlSec)
```

Comment:
Great question!

if we use `store, err := ...` here, it would be incorrect. Because then we actually created a new `store` and `err` variable, because we are in a new scope. And `store` would be a different variable than `var store master.Store`. What happened here is called `store` is "shadowed".

Please see here for me detail.
https://developmentality.wordpress.com/2014/03/03/go-gotcha-1-variable-shadowing-within-inner-scope-due-to-use-of-operator/

`go vet -shadow` can detect them:
https://golang.org/cmd/vet/#hdr-Shadowed_variables

`:=` is called [short variable declarations](https://golang.org/ref/spec#Short_variable_declarations)